core/src/main/java/org/bitcoinj/crypto/HDPath.java

```diff
@@ -172,13 +174,13 @@ public static HDPath m(ChildNumber... children) {
      */
     public static HDPath parsePath(@Nonnull String path) {
         List<String> parsedNodes = SEPARATOR_SPLITTER.splitToList(path);
-        boolean hasPrivateKey = false;
-        if (!parsedNodes.isEmpty()) {
-            final String firstNode = parsedNodes.get(0);
-            if (firstNode.equals(Character.toString(PREFIX_PRIVATE)))
-                hasPrivateKey = true;
-            if (hasPrivateKey || firstNode.equals(Character.toString(PREFIX_PUBLIC)))
-                parsedNodes.remove(0);
+
```

Comment:
Honestly, I think this is a perfect example where an early exit would be much more readable:

`if (parsedNodes.isEmpty()) return new HDPath(false, Collections.emptyList());`

…and the entire rest of the method wouldn't need to deal with this case.

Reply:
By the way, maybe our `splitToList()` guarantees a non-empty list? I vaguely remember that sometimes `"".split()` returns `[ "" ]`. But this would need to be documented first, and maybe test cases added if not already present.

Reply:
I'm rethinking the approach further and going in a slightly different direction. Starting with: https://github.com/bitcoinj/bitcoinj/pull/3685

Reply:
And we should definitely add more test cases. Note also, that 1b3863c8376e0ada3ffd87cae2e7736fbfeebe80 allow us to add tests at a lower level.